File: leetcode-75/10-binary-tree-bfs/39-binary-tree-right-side-view/solution.ts
/**
 * LeetCode 199. Binary Tree Right Side View
 * https://leetcode.com/problems/binary-tree-right-side-view/
 *
 * Approach: BFS level order traversal. For each level, keep only the
 * value of the last node dequeued (the rightmost node at that depth).
 *
 * Time:  O(n) - every node is visited once
 * Space: O(n) - queue holds up to the widest level; output holds one
 *               value per level
 */

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function rightSideView(root: TreeNode | null): number[] {
  const result: number[] = [];
  if (root === null) return result;

  const queue: TreeNode[] = [root];

  while (queue.length > 0) {
    const levelSize = queue.length;
    let rightmostValue = 0;

    for (let i = 0; i < levelSize; i++) {
      const node = queue.shift() as TreeNode;
      rightmostValue = node.val;

      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right);
    }

    result.push(rightmostValue);
  }

  return result;
}
if (require.main === module) {
  // Example usage:
  const root = new TreeNode(1, new TreeNode(2, null, new TreeNode(5)), new TreeNode(3, null, new TreeNode(4)));
  console.log(rightSideView(root)); // [1, 3, 4]
}

export { rightSideView, TreeNode };

File: leetcode-75/10-binary-tree-bfs/39-binary-tree-right-side-view/solution.py
"""
LeetCode 199. Binary Tree Right Side View
https://leetcode.com/problems/binary-tree-right-side-view/

Approach: BFS level order traversal. For each level, keep only the
value of the last node dequeued (the rightmost node at that depth).

Time:  O(n) - every node is visited once
Space: O(n) - queue holds up to the widest level; output holds one
              value per level
"""

from collections import deque
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if root is None:
            return []

        result = []
        queue = deque([root])

        while queue:
            level_size = len(queue)
            rightmost_value = 0

            for _ in range(level_size):
                node = queue.popleft()
                rightmost_value = node.val

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            result.append(rightmost_value)

        return result


if __name__ == "__main__":
    sol = Solution()
    root = TreeNode(1, TreeNode(2, None, TreeNode(5)), TreeNode(3, None, TreeNode(4)))
    print(sol.rightSideView(root))  # [1, 3, 4]

    root2 = TreeNode(1, TreeNode(2, TreeNode(4, TreeNode(5))), TreeNode(3))
    print(sol.rightSideView(root2))  # [1, 3, 4, 5]
